src/Amadevus.RecordGenerator.Generators/PartialGenerator.cs

```diff
@@ -0,0 +1,50 @@
+﻿using System;
+using System.Linq;
+
+namespace Amadevus.RecordGenerator.Generators
+{
+    using System.Collections.Generic;
+    using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+    internal interface IPartialGenerator
+    {
+        Generation Generate(RecordDescriptor descriptor, Features features);
+    }
+
+    static class PartialGenerator
+    {
+        public static IPartialGenerator Create(Func<RecordDescriptor, Features, Generation> generator) =>
+            new DelegatingPartialGenerator(generator);
+
+        public static IPartialGenerator Create(Features implementedFeatures, Func<RecordDescriptor, Features, Generation> generator) =>
+            Create(generator).IntersectFeatures(implementedFeatures);
+
+        public static IPartialGenerator Create(Features implementedFeatures, Func<RecordDescriptor, Generation> generator) =>
+            Create(implementedFeatures, (descriptor, _) => generator(descriptor));
+
+        public static IPartialGenerator IntersectFeatures(this IPartialGenerator generator, Features features) =>
+            Create((rd, fs) => (fs & features) != Features.None
+                             ? generator.Generate(rd, fs)
+                             : null);
+
+        public static IPartialGenerator Member(Features features, Func<RecordDescriptor, MemberDeclarationSyntax> member) =>
+            Create(features, descriptor => Generation.Empty.AddMember(member(descriptor)));
+
+        public static IPartialGenerator Combine(params IPartialGenerator[] generators) =>
```

Comment:
> Maybe we could make `IPartialGenerator`/`GenerationResult` ?

Not sure I get you here.

> A very random thought after reading this class. I don't know how, or from what exactly, but aggregating this, intersecting, etc. looks similar to LINQ.

Your point being? It's not surprising that it looks similar to LINQ because both use the spirit of functional composition.

Reply:
Well obviously my first sentence was not grammatically correct anyway, I think I lost some words on the brain-hand processing.

What I mean is making it LINQ-compatible via implementing IEnumerable, Select, Where, etc but again I don't really know how or why. Just an idea.

```csharp
from generator in Generators
from result in generator.Make(recordDescriptor)
group result by result.Modifiers into g
from item in g
select ...
```

It makes no sense. Let's end this conversation 😅

Reply:
> What I mean is making it LINQ-compatible via implementing IEnumerable, Select, Where, etc but again I don't really know how or why. Just an idea.

This is a complex idea and it would be _academically_ interesting to see if one could express generation with monadic composition but I think the energy spent doing that wouldn't justify the end. After all, this is a project with a small scope. We should focus on _some_ internal design (this PR) and more on getting records right and going 1.0 before C# records hit the street. So, yeah, let's move along. :)